### PR TITLE
Support sequence of losses in backwards pass

### PR DESCRIPTION
**Pull request by Landanjs**
For the non-deepspeed sections in trainer.py, it seems like the trainer aims to handle models that return a tuple of losses. One issue is that the backwards pass is called on each individual loss which will cause the error:

```
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: Trying to backward through the graph a second time (or directly access saved tensors after they have already been freed). Saved intermediate values of the graph are freed when you call .backward() or autograd.grad(). Specify retain_graph=True if you need to backward through the graph a second time or if you need to access saved tensors after calling backward.
```

In practice, most people sum multiple losses, then call backwards on the summed losses. This PR aims to do just that. This allows models that return a tuple of losses to be trained and will make it easier to log multiple losses in the future

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -1814,8 +1814,10 @@ def _train_microbatch(self, use_grad_scaling: bool, current_batch_size: int, tot
                     loss.mul_(microbatch_num_samples / current_batch_size)
                     total_loss += loss.detach().clone()
             else:
+                final_loss = self._device.tensor_to_device(torch.zeros(size=(1,)))
```
Comment (ravi-mosaicml) [inline composer/trainer/trainer.py:1817]: Do we need to worry about the dtype of loss? Or will torch convert the dtype if necessarry?

**Review by ravi-mosaicml — APPROVED** [review]
LGTM; If I remember my calculus I think this should be equivalent?

File: composer/trainer/trainer.py (hunk 1 above)
Comment (Landanjs) [reply composer/trainer/trainer.py:1817]: I think it should be fine? I did this because `total_loss` is set this way, should I just do `final_loss = 0`?

Comment (ravi-mosaicml) [reply composer/trainer/trainer.py:1811]: Just manually tested it, and torch keeps the more precise dtype, so this should be fine.

**Pull request merged**